### ISSUE TITLE
`beartype`d modules with forward references can't be `importlib.reload`ed

### ISSUE THREAD

**Issue by posita** [beartype/beartype#49]
This is probably a _**super**_ esoteric use case, but I stumbled across this (quite on accident):

``` python
# karma.py

from __future__ import annotations
from beartype import beartype

class Chameleon:
    def __init__(self, colors: str):
        self.colors = colors
    @classmethod
    @beartype
    def like_my_dreams(cls) -> Chameleon:
        return Chameleon("red, gold, and green")
```

``` python
# test_karma.py

import unittest

class TestReloadBearTyped(unittest.TestCase):
    def test_1_imported(self) -> None:
        import karma
        assert karma.Chameleon.like_my_dreams().colors == "red, gold, and green"

    def test_2_reloaded(self) -> None:
        import karma
        from importlib import reload
        karma = reload(karma)  # <-- this pulls the rug out from under the bear (comment this out and all tests pass)
        assert karma.Chameleon.like_my_dreams().colors == "red, gold, and green"

    def test_3_imported_again(self) -> None:
        self.test_1_imported()  # <-- once the beartyped module is reloaded in the prior test, this fails too

if __name__ == "__main__":
    unittest.main()
```

Results:

``` sh
% python test_karma.py
.EE
======================================================================
ERROR: test_2_reloaded (__main__.TestReloadBearTyped)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/…/test_karma.py", line 14, in test_2_reloaded
    assert karma.Chameleon.like_my_dreams().colors == "red, gold, and green"
  File "<string>", line 18, in like_my_dreams
  File "/…/lib/python3.9/site-packages/beartype/_decor/_error/errormain.py", line 316, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintPepReturnException: @beartyped Chameleon.like_my_dreams() return <karma.Chameleon object at 0x105c30eb0> violates type hint <class 'karma.Chameleon'>, as <karma.Chameleon object at 0x105c30eb0> not <class "karma.Chameleon">.

======================================================================
ERROR: test_3_imported_again (__main__.TestReloadBearTyped)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/…/test_karma.py", line 17, in test_3_imported_again
    self.test_1_imported()
  File "/…/test_karma.py", line 8, in test_1_imported
    assert karma.Chameleon.like_my_dreams().colors == "red, gold, and green"
  File "<string>", line 18, in like_my_dreams
  File "/…/lib/python3.9/site-packages/beartype/_decor/_error/errormain.py", line 316, in raise_pep_call_exception
    raise exception_cls(  # type: ignore[misc]
beartype.roar.BeartypeCallHintPepReturnException: @beartyped Chameleon.like_my_dreams() return <karma.Chameleon object at 0x105c30ee0> violates type hint <class 'karma.Chameleon'>, as <karma.Chameleon object at 0x105c30ee0> not <class "karma.Chameleon">.

----------------------------------------------------------------------
Ran 3 tests in 0.043s

FAILED (errors=2)
```

[![Juggling Bear](https://images.squarespace-cdn.com/content/v1/57432be37da24f8f64d0b7d4/1505329420481-HGY3I4NVFGGZ2CF7P5CE/juggling_bear.jpg?format=256w)](https://www.scottgustafson.com/misc-animals-studio-prints)

**Comment by posita** [comment]
Oh, interesting. Reloading only fails with forward references. This works:

``` python
# karma.py

from beartype import beartype

class Chameleon:
    def __init__(self, colors: str):
        self.colors = colors
    @classmethod
    @beartype
    def like_my_dreams(cls) -> "Chameleon":  # <-- string-ified return type
        return Chameleon("red, gold, and green")
```

``` sh
% python test_karma.py
...
----------------------------------------------------------------------
Ran 3 tests in 0.057s

OK
```

**Comment by leycec** [comment]
*Dancing jester bear shrieks with panic.* After calming frayed nerves with soothing honey, dancing jester bear (hereafter, DJB) thanks @posita profusely for this exciting edge case that tunnels into the very bowels of sanity itself.

DJB admits he never even considered reloading modules containing decorated callables annotated with forward references via `importlib.reload()`. Despite the gnawing suspicion that @posita is now delving into the dark arts of dynamic plugin systems for [`dyce`](https://github.com/posita/dyce), DJB reluctantly agrees that this does indeed signal bad stuff for `beartype`. :sweat: 

Let's patch this up and kick out `beartype 0.9.0`, DJB!

<sup>*this has been a message from the autumnal steppes of Canada*</sup>

**Comment by posita** [comment]
To be fair, I have absolutely zero use cases for this. I would assume it enjoys about as much presence in the real world as **_actual_** DJBs. (FWIW, I discovered it by casual debugging/tinkering around with something in IPython while being too lazy to restart the session as I made changes to the underlying code. I wouldn't exactly call that a functional workflow). I am beyond impressed that you even have an idea of how to fix it.

**Comment by leycec** [comment]
> I am beyond impressed that you even have an idea of how to fix it.

...*but I wish I didn't.* Some issues bring you so close to the edge of darkness, you'd swear they were staring back at you with beady rheumy eyes. The issue is this:

```
from __future__ import annotations
```

Because the Future is a hip and cool place to be, that import conditionally enables [PEP 563](https://www.python.org/dev/peps/pep-0563) support... *from the future.* This means that CPython will **unparse** all annotation objects into strings across the current module. Uncoincidentally, these strings are semantically and syntactically identical to **forward references.** A picture of doom soon emerges. Thus:

```
# CPython implicitly unparses this pre-parsed code...
from __future__ import annotations
from beartype import beartype

class Chameleon:
    def __init__(self, colors: 'str'):
        self.colors = colors
    @classmethod
    @beartype
    def like_my_dreams(cls) -> Chameleon:
        return Chameleon("red, gold, and green")

# ...into this post-unparsed code, which is what @beartype sees.
from __future__ import annotations
from beartype import beartype

class Chameleon:
    def __init__(self, colors: 'str'):
        self.colors = colors
    @classmethod
    @beartype
    def like_my_dreams(cls) -> 'Chameleon':
        return Chameleon("red, gold, and green")
```

So, `def __init__(self, colors: str)` becomes `def __init__(self, colors: 'str')` *and* `def like_my_dreams(cls) -> Chameleon:` becomes `def like_my_dreams(cls) -> 'Chameleon':`. So, `@beartype` sees a string `'Chameleon'` rather than the class object `Chameleon`, which `@beartype` then resolves back into the original class object `Chameleon` by:

1. First, inspecting the global scope of the `like_my_dreams()` method for an attribute named `Chameleon`. If that fails...
2. Next, [**dynamically crawling up the call stack from the decorated method `like_my_dreams()` until finding a stack frame defining `Chameleon` as a local attribute.**](https://github.com/beartype/beartype/blob/ad20c2c3bd2590331ee3f3de702371d86ec8cb55/beartype/_util/func/utilfuncscope.py#L190)

`@beartype` actually has [the best PEP 563 resolution algorithm on the market](https://github.com/beartype/beartype/blob/main/beartype/_decor/_pep563.py). I know this, because I really wanted to steal someone else's rather than implement my own. Turns out: *no one implements their own.* Everyone just calls `typing.get_type_hints()`, which is exactly the wrong thing to do, because `typing.get_type_hints()` is functionally brain dead and can't actually resolve commonplace things like nested locals.

Since Python ≥ 3.10 unconditionally enables PEP 563, this means that every *other* runtime type checker (e.g., pydantic, typical, typeguard) breaks under Python ≥ 3.10 when confronted with code that used to run under older Python versions. Meanwhile, `@beartype` has no such issues. We preserve forward compatibility, because we ignored `typing.get_type_hints()` and instead did everything ourselves and made sure everything still worked under all edge cases.

I may still be salty about the whole thing. I'm pretty sure my cheeks are permanently cracked with the salty exudate of the boiling frustration I felt while implementing all of this. :joy: 

...*which brings us back to you.* Yes, you – you immeasurable ponderous genius! Because resolving PEP 563-postoned annotations has worst-case complexity `O(n)` for `n` the distance in call stack frames between the call of the `@beartype()` decorator to the call of the top-most parent scope transitively declaring the decorated callable in its submodule, we cache the results so we don't have to do it every time we see that callable. Clearly, **our caching sucks.**

And that's the core underlying issue. I probably "neglected" to unit test that our caching of resolved PEP 563-postoned annotations actually worked. In my defense, I was tired. I'm still tired, but now disappointed with my prior self. I'm now bludgeoning the air in a pantomime of an angry cat confronted with an empty litter box containing no cat litter.

Curse you, younger @leycec! Why you make older @leycec do your work for you? `</sigh>`

**Comment by leycec** [comment]
**Woops.** Just realized I incoherently lore-dumped there for a bit. At least Google now has something to index when type devs search for "pep 593 y u so bad."

*tl;dr* — let's fix this and pretend this never happened. :wink:

**Comment by posita** [comment]
Sorry for not responding earlier! (As you likely know, I have my own gripes about "unfinished business" in the Python standard library.) No apologies necessary for lore-dumping. Thanks for making an esoteric and opaque topic accessible!

For those following along at home, _**this**_ is why you should use this product (beyond the awesome utility and the deliciously distinguished prose):

**_@leycec KNOWS. HIS. CRAFT._**

Much gratitude for letting the rest of us inside the brain machine. 💖

**Comment by leycec** [comment]
**Aww.** You so noice. Really – I'm just obsessive. But you're the Big Guy™ at DropBox! You probably even got to genuflect in obeisance before Our Breathing God of Guido, which is several orders of magnitude beyond the spectral limits of my resume, which mostly consists of "Live in wood cabin. Have cats and beautiful wife. Will grovel for money on camera."

Also, you do [dyce](https://github.com/posita/dyce) and are funnier and cleverer. I actually meant to ask you about the former, so let's go! Have you posted [dyce](https://github.com/posita/dyce) to social media yet? I know... *I know.* I prefer the introverted solace of my  man-hovel, too. But [/r/Python](https://reddit.com/r/Python), [/r/pystats](https://www.reddit.com/r/pystats), and [/r/rpg](https://www.reddit.com/r/rpg)  would all adore that scandalous D&D DSL you concocted. I mean, `2@d6`. Am I right? Of course I'm right. That's just hot.

Also, *wut is dis:*

:sparkling_heart: :sparkling_heart: :sparkling_heart: :sparkling_heart: :sparkling_heart:

Why didn't someone tell me about the sparkling heart emoji before now? (╯°□°)╯︵ ┻━┻

**Comment by leycec** [comment]
**Ho, ho, ho.** Merry post-Canadanian federal election and almost Canadian Thanksgiving! It's that bitterly cold time of the year again when the trees start dumping dangerous detritus  everywhere and I actually resolve a horrible issue. That means this issue.

So, it turns out `importlib.reload()` is [flawed by design and no one should use it, but everyone still does because it's there and sounds nice](https://stackoverflow.com/a/438845/2809027). The sanest alternative is [the `%autoreload` magic extension under (IPython|Jupyter Notebook|Jupyter Lab)](https://stackoverflow.com/a/49348585/2809027).

But... who cares! This was a fantastic puzzle to unravel. I thank you profusely, @posita, for feeding my mind's eye with suspicious candy. This is the last action item on my  `@beartype 0.9.0` release schedule, which I hope to excrete onto the world tomorrow.

`@beartype 0.9.0` fixes both this specific `importlib.reload()` scenario *and* another related scenario that looks superficially different but actually boils down to the same exact thing:

``` python
# This looks bad and is bad. But it's actually valid Python, so...
# What you gonna do?
from __future__ import annotations

MuhClass = "Just kidding! Had you going there, didn't I?"
class MuhClass:
    @beartype
    def muh_method(self) -> MuhClass:
        return MuhClass()
```

Given that code, `@beartype` previously raised a similar exception. Now, `@beartype` proudly flexes like a type-checking chad bear.

Note that `@beartype` will continue raising exceptions on callables annotated by **parent** type hints subscripted by **child** class self-references (e.g., `list[MuhClass]`). There's not much anyone can (or should) do about that, so it's a safer habit to *always* replace class self-references with forward references – regardless of whether they're parent or child type hints:

``` python
from __future__ import annotations

# Replace nested class self-references, which will continue
# failing at runtime...
class MuhClass:
    @beartype
    def muh_method(self) -> list[MuhClass]:
        return [MuhClass()] * 69


# ...with nested forward references, which will continue
# working at runtime...
class MuhClass:
    @beartype
    def muh_method(self) -> list['MuhClass']:
        return [MuhClass()] * 69
```

So, `def muh_method(self) -> list['MuhClass']:` rather than `def muh_method(self) -> list[MuhClass]:` is what I'm saying. *Thus spaketh Bearathustra.*

Lastly, I literally cannot believe you embedded [Boy George and the Culture Club](https://www.youtube.com/watch?v=JmcA9LIIXWw). I knew you'd go far; I just didn't know how far. For those who are about to weep into their lukewarm 80's pop culture slurpees, I salute you.

**Comment by posita** [comment]
_[… embarrassingly averts gaze for using `importlib.reload`, but silently mouths with fists clenched, "Never again."]_ 😒😊😁

I'm happy (😬?) to bring this up for the entertainment-as-an-investigative-puzzle value, but otherwise sad if this occupied your time without yielding enjoyment. 🎭 In any event, I am grateful for the education, as always. 🙏

—Eternally in service of Bearathustra, the source of all ursine truth and light.

Since I have nothing else of substance to add, I will attempt to distract from my lack of further contribution with my best attempt at an emoji chad bear….

```
     S'up, bearah?
    /
 🐻
💪👔🍻
 👖 
 🐾
```

**Comment by leycec** [comment]
> ...but otherwise sad if this occupied your time without yielding enjoyment.

Oh, no! It was great fun, actually. I was reminded nostalgically of dead-tree logic puzzle books I repetitiously solved like a child automaton in the fecund back seat of the family car during purgatorial car trip vacations across the vast featureless void of the American Midwest. Little did I know, but I was training then for a promising future in... *type-checking.* `</chuckles>`

> emoji chad bear….

**OMG.** ASCII art chad bear constructed like Voltron: Legendary Defender from the burned-out husks of long-forgotten emoji! This day too shall live in infamy.

 ` ...do you even flex honey brewskies, bearah?`
 `/`
 🐻
💪👔🍻
 👖 
 🐾

**Comment by posita** [comment]
> Oh, no! It was great fun, actually. I was reminded nostalgically of dead-tree logic puzzle books I repetitiously solved like a child automaton in the fecund back seat of the family car during purgatorial car trip vacations across the vast featureless void of the American Midwest. Little did I know, but I was training then for a promising future in... _type-checking._ `</chuckles>`

Can't … breathe … . 🤣 OMG, this is priceless. Tragic. But _**priceless**_.

**Comment by leycec** [comment]
I'll just [leave this here](https://github.com/leycec). :wink: :stuck_out_tongue_winking_eye:
